app/eth-watcher.hoon

```diff
@@ -0,0 +1,556 @@
+::  watcher: ethereum event log collector
+::
+/+  *eth-watcher
+::
+=,  ethereum
+=,  rpc
+::
+|%
+++  state
+  $:  eyes=(map name eye)
+  ==
+::
+++  eye
+  $:  config
+      latest-block=@ud
+      filter-id=@ud
+      poll-timer=@da
+      snapshot
+      sap=history
+  ==
+::
+++  history
+  $:  interval=_100
+      max-count=_10
+      count=@ud
+      latest-block=@ud
+      snaps=(qeu snapshot)
+  ==
+::
+++  move  (pair bone card)
+++  card
+  $%  [%hiss wire (unit user:eyre) mark %hiss hiss:eyre]
+      [%wait wire @da]
+      [%rest @da]
+      [%info wire desk nori:clay]
+      [%diff %eth-watcher-update update]
+      [%quit ~]
+  ==
+--
+::
+|_  [bowl:gall state]
+::
+++  prep
+  |=  old=(unit state)
+  ?~  old
+    [~ ..prep]
+  [~ ..prep(+<+ u.old)]
+::
+++  poke-noun
+  |=  [what=?(%debug %debug-init %save %load) =name]
+  ^-  (quip move _+>)
+  =+  eye=(fall (~(get by eyes) name) *eye)
+  ?-  what
+      %debug
+    ~&  [%log-lent (lent logs.eye)]
+    ~&  [%last-heard last-heard-block.eye]
+    [~ +>.$]
+  ::
+      %debug-init
+    =-  done:(init:watcher name -)
+    ^-  config
+    :*  (need (de-purl:html 'http://104.198.35.227:8545'))  ::  parity
```

Comment:
What are these IPs?

Reply:
Tlon-hosted Ethereum nodes. @ohAitch good Q, not entirely sure if they have domains pointing to them. Not sure if we want to either. I should clean out/up most of the debug stuff before merge.

Reply:
`.tlon.io` then :P

It seems pretty valuable for code readability; is the concern here making them harder to find and DoS? ~sometimes the person you end up making the system secure against… is yourself~

Reply:
https://blog.ethereum.org/2015/08/29/security-alert-insecurely-configured-geth-can-make-funds-remotely-accessible/

> ADVISORY: Never enable JSON-RPC interface on an internet-accessible machine without a firewall policy in place to block the JSON-RPC port (default: 8545).

Maybe things changed between now and then, but I generally feel uneasy about hosting a publicly accessible JsonRPC port for anything. Somewhere down the line there will be some security vuln. available, and at very least someone will be able to eg. serve fake data to other users relying on this host.

Reply:
I mean, they'll be able to serve _stale_ data, if a single ethereum node can serve you _false_ data about the blockchain we have bigger problems. Unless this app isn't checking signatures / timestamps / difficulty?

(though finding a self-hosted JSON-RPC wrapper that does claim to be safe for public consumption does seem like a good idea, having to reprovision the server in question sounds annoying at least)

Reply:
@ohAitch 

Worst case scenario. There's is a 0-day bug in parity/geth RPC, someone gains shell access to this host. They totally replace the geth/parity with their own version, that servers *any* data they desire.


RPC ports are just lot of API surface. A lot of commands, with varying functionality - accessing filesystems etc., and since it's recommended to keep it for localhost consumption, it doesn't get as much scrutiny.

Any software that only consumes RPC data and does not directly connect with p2p network, is at a mercy of the node they are querying giving them correct data, as there is nothing to cross-match it with.

Even if this host is not compromised by an outsider, all users consuming this RPC, are trusting Tlon administrators.

Reply:
@dpc that advisory was and still is about cases where you have wallets accessible/created via the RPC interface. Having wallets with *anything* attached to them (funds, permissions, whatever) connected to the internet is obviously dangerous and stupid. Exposing an RPC node just for reading the blockchain is fine and not uncommon.

@philipcmonk care to confirm my sanity here?

Reply:
We use http://eth-mainnet.urbit.org:8545/ for preboot validation.

That advisory doesn't apply, as @Fang- says, however @dpc 's concern about needing to trust the node to give valid information is valid.  This is a problem with *any* application that isn't running a full node inside (including metamask, mycrypto, augur, etc).  They can't steal your money directly, but they can fool you into thinking the blockchain says something other than what it does.

We should of course switch to HTTPS -- that would reduce the MITM attack surface.  But if the server the node is on is compromised, then all bets are off.  Each user should decide for themselves what level of security do they want.  They can run their own node(s) and point this at them, or they could use Infura if they trust them (if we had a websocket client, or else getFilterChanges doesn't work).  For now, Tlon hosts a couple nodes, and I think that's fine for a default.  It wouldn't be crazy to find several sources and verify they're all giving the same information.  This is definitely worth revisiting some time.

Reply:
Note that the addresses of Tlon's nodes aren't in this app as defaults, but as hardcoded values for a hardcoded config that's used during debugging. In actual use, the user needs to decide which node address to pass in, which is where the considerations mentioned above come in.

Reply:
I'm taking that as a yes on `this app isn't checking signatures / timestamps / difficulty`: we're speaking over RPC to Tlon software that claims to be tracking the Azimuth contract
Though if it's only doing so for debugging that does sound fine.

Back to the original "what are there IPs" question: any reason not to DNS them as `test-eth-parity.urbit.org` etc?

Reply:
The parity one is DNSed to eth-mainnet.urbit.org.  But yeah, the user should be getting their own node from somewhere.

@ohAitch that's correct, we don't implement a light node in this app.  Even if we did, that still wouldn't guarantee canonicality unless we could trust its peers.  Even a full node is only as good as its peers.  If its peers disagree, then it can choose the right one by checking difficulty, but eg if it only has one peer it won't have the right info.  It needs to have at least one honest peer.